### PR TITLE
Add standard date functions and (optional) MaintenanceWindow to RDS

### PR DESCRIPTION
**Pull request by ken9ross**
## Intent of Change

- New feature (non-breaking change which adds functionality)

## Description
Standard date manipulation functions added
Standard MaintenanceWindow child added to RDS

## Motivation and Context
Precondition for PR to address https://github.com/hamlet-io/engine/issues/159

## How Has This Been Tested?
Locally

### Prerequisite PRs:

- None

### Dependent PRs:

- TBA

### Consumer Actions:

- Functionality to be supported within provider plugins for MaintenanceWindow

### CODE REVIEW

File: engine/common.ftl
Comment (roleyfoley) [inline engine/common.ftl:51]: What are these values for with the current date? Could we instead use ?now which gets the current time?

File: providers/shared/components/cache/id.ftl
Comment (roleyfoley) [inline providers/shared/components/cache/id.ftl:74]: Should we default the TimeZone to UTC to make it easier to follow?

File: engine/common.ftl
Comment (ml019) [inline engine/common.ftl:1]: If generic, these functions should be in base.ftl rather than common.ftl.

File: providers/shared/components/cache/id.ftl
Comment (ml019) [inline providers/shared/components/cache/id.ftl:52]: Feels like this should be an attribute set given it is turning up in multiple places.

**Review by ml019 — DISMISSED** [review]

**Comment by ml019** [comment]
@ken9ross a few suggestions...

File: engine/common.ftl
Comment (ken9ross) [reply engine/common.ftl:51]: These are to convert a "day" to a date which is sufficiently far away from any leap years etc to allow for timezone manipulations and date arithmetic (eg Saturday 01:00AEST to become Friday 15:00UTC)

File: providers/shared/components/cache/id.ftl
Comment (ken9ross) [reply providers/shared/components/cache/id.ftl:74]: Maybe I configured it incorrectly, but if I set a default on the children o, I was then explicitly setting the maintenance window and couldn't leave it to AWS to configure an "appropriate" window (comment mostly applies to RDS which use the region to determine "appropriate")

File: engine/common.ftl
Comment (ken9ross) [reply engine/common.ftl]: Cool

File: engine/common.ftl
Comment (ken9ross) [reply engine/common.ftl]: Done

File: providers/shared/components/cache/id.ftl
Comment (ken9ross) [reply providers/shared/components/cache/id.ftl:52]: Done but left the queuehost which has a default as an explicit definition

File: engine/common.ftl
Comment (roleyfoley) [reply engine/common.ftl]: I think we need a comment on this as to why we have added fixed dates, its a bit weird to see specific dates in the code is all.

File: providers/shared/components/cache/id.ftl
Comment (roleyfoley) [reply providers/shared/components/cache/id.ftl]: Thats the isConfigured check on the Maintenance window right? I.e. if someone has configured a maintenance window the defaut TimeZone is to use UTC.

File: providers/shared/components/cache/id.ftl
Comment (ml019) [reply providers/shared/components/cache/id.ftl:52]: You can override the default attributeset with a default value in the queuehost.

File: engine/common.ftl
Comment (ken9ross) [reply engine/common.ftl]: Cool, done

File: providers/shared/components/cache/id.ftl
Comment (ken9ross) [reply providers/shared/components/cache/id.ftl]: I don't see a function isConfigured. Is it in master? I have a isLayerConfigured and a isConfigurationIncluded but neither seems to be what you're talking about

Also I didn't think there was a difference between something set in a solution file, inherited from a shared solution file or set by a default, when it came to processing. Am I missing something?

Comment (ml019) [reply providers/shared/components/cache/id.ftl]: IsConfigured is added at run-time to say whether the parent has explicitly been defined. Do a global search in the aws provider and you'll see the usage.

Comment (ken9ross) [reply providers/shared/components/cache/id.ftl]: The function is isPresent and the property is .Configured (had a chat with Roley at the standup). I'm leaning towards .Configured, because isPresent relies on an Enabled:true as well and you can't disable maintenance windows, so I think that'd be misleading

The more I think about this, the more I'm in favour of making the day, time and timezone mandatory with no defaults rather than having someone specify to do maintenance on Wednesday at 01:00 and have it trigger at 11:00 due to not specifying the timezone and picking up the UTC default. Thoughts?

If a child attribute has Mandatory:true does that force the existence of the parent, or is the child attribute only mandatory if the parent is defined? ie Can I enforce using Mandatory:true or do I need to explicitly code a fatal error for a partial maintenance window specification?

Comment (roleyfoley) [reply providers/shared/components/cache/id.ftl]: Mandatory applies to the whole configuration object not just a child configuration. So if the property is conditional then you need to handle it outside of the configuration process

Comment (ken9ross) [reply providers/shared/components/cache/id.ftl]: Cool, added standard function testMaintenanceWindow to handle including Azure v AWS options based on current preview

**Comment by ken9ross** [comment]
This looks like something's gone screwy with the rebase. @roleyfoley Can I beg for assistance to figure out what's happened here?

**Comment by roleyfoley** [comment]
@ken9ross The branch needed to be rebased to realign the changes made in the master branch and this branch

**Review by roleyfoley — APPROVED** [review]

**Comment by roleyfoley** [comment]
@ml019 could you please approve this one if you are happy with it. Since you asked for changes you need to say they are done

**Comment by ken9ross** [comment]
@ml019 Updated to the conventional commit syntax

**Review by roleyfoley — APPROVED** [review]

**Pull request merged**
